### PR TITLE
Upgrade jackson from 2.5.3 to 2.8.4

### PR DESCRIPTION
**Pull request by yuesong**

### CODE REVIEW

**Comment by jhorwit2** [comment]
Curious, does this work with jackson 2.7.x also?

**Comment by stevedonnelly** [comment]
Changes all appear to be reasonable (to someone who isnt an expert on the jackson API).

Josh, it's not 100% compatible with 2.7.x, although it appears to be just a two line change to make it so if needed.  Hopefully your other library doesn't depend on any jackson internals and can go from 2.7 to 2.8?

**Comment by jhorwit2** [comment]
AFAIK everyone on dw9 can use this version of jackson. If not we can just make a custom release branch for jackson 2.7 to address it.

File: src/main/java/com/addthis/codec/jackson/CodecTypeDeserializer.java [hunk 1]
```diff
@@ -125,7 +125,9 @@ public Object deserializeTypedFromAny(JsonParser jp, DeserializationContext ctxt
             if (bean != null) {
                 return bean;
             } else {
-                JsonDeserializer<Object> deser = _findDeserializer(ctxt, null);
+                // Jackson 2.6+ throws NPE on null typeId parameter (underlying Map changed from HashMap
+                // to ConcurrentHashMap), so use empty string instead of null
+                JsonDeserializer<Object> deser = _findDeserializer(ctxt, "");
```
Comment (tea-dragon) [inline src/main/java/com/addthis/codec/jackson/CodecTypeDeserializer.java]: Pretty sure that statement always resolved into getting the deserializer for the default impl, and we can just do that directly.

**Pull request merged**